# DS3000 Day 11 
## (Day 10 was Lab 3)

Oct 18 2022

Admin:
- Qwickly attendance
- due Tonight before midnight
    - hw4
- on Friday: Project Teams finalized
- also on Friday: Quiz 2 will be released, will have weekend to complete it
    - **Do not forget this time; any quizzes not submitted by Sunday at 11:59 pm will be given 0**
- hw5 will be posted by the end of today, will have a one week to do it (due: Oct 25)

Content:
- Cleaning Data

## DataFrame Manipulations
- Indexing
    - set_index() / reset_index()
    - hierarchical indexing
- renaming columns
- copying vs modifying DataFrame `inplace`
- Sorting rows / cols of DataFrame
    - DataFrame.sort() 
        - by index or column name
    - DataFrame.sort_values()
        - by the values themselves

# Index of a DataFrame
- index of a DataFrame is name of the row
- index need not be unique
    - ... under the hood pandas tracks the position of the row rather in addition to its name

## DataFrame.reset_index()
* "resets" current index:
    * moves current index into a column
    * creates new index as integers

In [1]:
import pandas as pd

df_poke = pd.read_csv('pokedata.csv', index_col='Pokedex')
df_poke.head()

Pokemon MainType SecondaryType  Height  Weight  Damage  BaseSpeed  \
Pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         Attack  Defense  
Pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [2]:
df_poke.index

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            142, 143, 144, 145, 146, 147, 148, 149, 150, 151],
           dtype='int64', name='Pokedex', length=151)

In [3]:
df_poke1 = df_poke.reset_index()
df_poke1.head()

Pokedex     Pokemon MainType SecondaryType  Height  Weight  Damage  \
0        1   Bulbasaur    Grass        Poison      24    25.2      45   
1        2     Ivysaur    Grass        Poison      39    28.7      60   
2        3    Venusaur    Grass        Poison      79   220.5      80   
3        4  Charmander     Fire           NaN      24    18.7      39   
4        5  Charmeleon     Fire           NaN      43    41.9      58   

   BaseSpeed  Attack  Defense  
0         45      49       49  
1         60      62       63  
2         80      82       83  
3         65      52       43  
4         80      64       58

In [4]:
df_poke1.index

RangeIndex(start=0, stop=151, step=1)

## DataFrame.set_index()
- .set_index() moves a column into the index position
    - existing index is discarded
        - want to keep this index as a column?
            - `.reset_index()` first

In [5]:
import pandas as pd 

df_poke = pd.read_csv('pokedata.csv')
df_poke.head()

Pokedex     Pokemon MainType SecondaryType  Height  Weight  Damage  \
0        1   Bulbasaur    Grass        Poison      24    25.2      45   
1        2     Ivysaur    Grass        Poison      39    28.7      60   
2        3    Venusaur    Grass        Poison      79   220.5      80   
3        4  Charmander     Fire           NaN      24    18.7      39   
4        5  Charmeleon     Fire           NaN      43    41.9      58   

   BaseSpeed  Attack  Defense  
0         45      49       49  
1         60      62       63  
2         80      82       83  
3         65      52       43  
4         80      64       58

In [6]:
df_poke.index

RangeIndex(start=0, stop=151, step=1)

In [7]:
df_poke1 = df_poke.set_index('Pokemon')
df_poke1.head()

Pokedex MainType SecondaryType  Height  Weight  Damage  BaseSpeed  \
Pokemon                                                                         
Bulbasaur         1    Grass        Poison      24    25.2      45         45   
Ivysaur           2    Grass        Poison      39    28.7      60         60   
Venusaur          3    Grass        Poison      79   220.5      80         80   
Charmander        4     Fire           NaN      24    18.7      39         65   
Charmeleon        5     Fire           NaN      43    41.9      58         80   

            Attack  Defense  
Pokemon                      
Bulbasaur       49       49  
Ivysaur         62       63  
Venusaur        82       83  
Charmander      52       43  
Charmeleon      64       58

In [8]:
df_poke1.index

Index(['Bulbasaur', 'Ivysaur', 'Venusaur', 'Charmander', 'Charmeleon',
       'Charizard', 'Squirtle', 'Wartortle', 'Blastoise', 'Caterpie',
       ...
       'Aerodactyl', 'Snorlax', 'Articuno', 'Zapdos', 'Moltres', 'Dratini',
       'Dragonair', 'Dragonite', 'Mewtwo', 'Mew'],
      dtype='object', name='Pokemon', length=151)

In [9]:
# note: we lose the previous index!
df_poke2 = df_poke1.set_index('Pokedex')
df_poke2.head()

MainType SecondaryType  Height  Weight  Damage  BaseSpeed  Attack  \
Pokedex                                                                     
1          Grass        Poison      24    25.2      45         45      49   
2          Grass        Poison      39    28.7      60         60      62   
3          Grass        Poison      79   220.5      80         80      82   
4           Fire           NaN      24    18.7      39         65      52   
5           Fire           NaN      43    41.9      58         80      64   

         Defense  
Pokedex           
1             49  
2             63  
3             83  
4             43  
5             58

In [10]:
# index need not be unique ...
df_poke3 = df_poke.set_index('MainType')
df_poke3.head()

Pokedex     Pokemon SecondaryType  Height  Weight  Damage  \
MainType                                                              
Grass           1   Bulbasaur        Poison      24    25.2      45   
Grass           2     Ivysaur        Poison      39    28.7      60   
Grass           3    Venusaur        Poison      79   220.5      80   
Fire            4  Charmander           NaN      24    18.7      39   
Fire            5  Charmeleon           NaN      43    41.9      58   

          BaseSpeed  Attack  Defense  
MainType                              
Grass            45      49       49  
Grass            60      62       63  
Grass            80      82       83  
Fire             65      52       43  
Fire             80      64       58

In [11]:
# non unique index has some funny behavior ...
# kind of like boolean indexing, right?  we can select all rows with 'MainType' = 'Water'
df_poke3.loc['Water', :].head()

Pokedex    Pokemon SecondaryType  Height  Weight  Damage  BaseSpeed  \
MainType                                                                        
Water           7   Squirtle           NaN      20    19.8      45         43   
Water           8  Wartortle           NaN      39    49.6      59         58   
Water           9  Blastoise           NaN      63   188.5      79         78   
Water          54    Psyduck           NaN      31    43.2      50         55   
Water          55    Golduck           NaN      67   168.9      80         85   

          Attack  Defense  
MainType                   
Water         48       65  
Water         63       80  
Water         83      100  
Water         52       48  
Water         82       78

## Hierarchical Indexing
You can use multiple columns as an index by passing a list to `.set_index()`

In [12]:
df_poke = pd.read_csv('pokedata.csv')
df_poke.head()

Pokedex     Pokemon MainType SecondaryType  Height  Weight  Damage  \
0        1   Bulbasaur    Grass        Poison      24    25.2      45   
1        2     Ivysaur    Grass        Poison      39    28.7      60   
2        3    Venusaur    Grass        Poison      79   220.5      80   
3        4  Charmander     Fire           NaN      24    18.7      39   
4        5  Charmeleon     Fire           NaN      43    41.9      58   

   BaseSpeed  Attack  Defense  
0         45      49       49  
1         60      62       63  
2         80      82       83  
3         65      52       43  
4         80      64       58

In [13]:
# set a hierarchical index (must pass list, not a tuple)
df_poke0 = df_poke.set_index(['MainType', 'Pokedex'])
df_poke0.head()

Pokemon SecondaryType  Height  Weight  Damage  BaseSpeed  \
MainType Pokedex                                                                
Grass    1         Bulbasaur        Poison      24    25.2      45         45   
         2           Ivysaur        Poison      39    28.7      60         60   
         3          Venusaur        Poison      79   220.5      80         80   
Fire     4        Charmander           NaN      24    18.7      39         65   
         5        Charmeleon           NaN      43    41.9      58         80   

                  Attack  Defense  
MainType Pokedex                   
Grass    1            49       49  
         2            62       63  
         3            82       83  
Fire     4            52       43  
         5            64       58

In [14]:
# accessing a hierarchical index
df_poke0.loc[('Water', 7), :]
#df_poke0.loc[('Water', 8), :]
#df_poke0.loc[('Water', 6), :] #be careful that you call something that exists

Pokemon          Squirtle
SecondaryType         NaN
Height                 20
Weight               19.8
Damage                 45
BaseSpeed              43
Attack                 48
Defense                65
Name: (Water, 7), dtype: object

In [15]:
# we can access all rows of hierarchy which match the first index
df_poke0.loc['Water', :].head()

Pokemon SecondaryType  Height  Weight  Damage  BaseSpeed  Attack  \
Pokedex                                                                       
7         Squirtle           NaN      20    19.8      45         43      48   
8        Wartortle           NaN      39    49.6      59         58      63   
9        Blastoise           NaN      63   188.5      79         78      83   
54         Psyduck           NaN      31    43.2      50         55      52   
55         Golduck           NaN      67   168.9      80         85      82   

         Defense  
Pokedex           
7             65  
8             80  
9            100  
54            48  
55            78

In [16]:
# but we must specify our indices in order
# that is, we can't get all the rows which match the second index
# without specifying the first
# df_poke0.loc[7, :]

### DataFrame.rename()
- gives new names to index (axis=0) or columns (axis=1)
    - passing a dict:
        - keys are old names
        - values are new names
    - passing a function:
        - applies to all names

In [17]:
df_poke = pd.read_csv('pokedata.csv', index_col = 'Pokedex')
df_poke.head()

Pokemon MainType SecondaryType  Height  Weight  Damage  BaseSpeed  \
Pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         Attack  Defense  
Pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [18]:
# we can pass a dict whose keys are old names and values are new names
dict_rename = {'MainType': 'type1',
               'SecondaryType': 'type2'}
df_poke.rename(dict_rename, axis=1).head()

Pokemon  type1   type2  Height  Weight  Damage  BaseSpeed  Attack  \
Pokedex                                                                         
1         Bulbasaur  Grass  Poison      24    25.2      45         45      49   
2           Ivysaur  Grass  Poison      39    28.7      60         60      62   
3          Venusaur  Grass  Poison      79   220.5      80         80      82   
4        Charmander   Fire     NaN      24    18.7      39         65      52   
5        Charmeleon   Fire     NaN      43    41.9      58         80      64   

         Defense  
Pokedex           
1             49  
2             63  
3             83  
4             43  
5             58

In [19]:
# we can also pass a function which operates on each index / column
def square_it(x):
    return x ** 2

df_poke.rename(square_it, axis=0).head()

Pokemon MainType SecondaryType  Height  Weight  Damage  BaseSpeed  \
Pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
4           Ivysaur    Grass        Poison      39    28.7      60         60   
9          Venusaur    Grass        Poison      79   220.5      80         80   
16       Charmander     Fire           NaN      24    18.7      39         65   
25       Charmeleon     Fire           NaN      43    41.9      58         80   

         Attack  Defense  
Pokedex                   
1            49       49  
4            62       63  
9            82       83  
16           52       43  
25           64       58

rename only renames index (axis=0) or column names (axis=1)

In [20]:
# we can also pass a function which operates on each index / column
def make_lower(x):
    return str(x).lower()

df_poke.rename(make_lower, axis=1).head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
Pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         attack  defense  
Pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [21]:
# notice that these are temporary changes; we would need to overwrite the object to enact them
df_poke.head()

Pokemon MainType SecondaryType  Height  Weight  Damage  BaseSpeed  \
Pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         Attack  Defense  
Pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

## sorting

- DataFrame.sort_index()
    - sorts rows / columns by the values of the index / column names
- DataFrame.sort_values()
    - sorts rows / columns by the values inside the dataframe
- `ascending` param:
    - if True then sorts in increasing order (default)
- `axis` param:
    - 0: sort rows
    - 1: sort columns

In [22]:
df_poke = pd.read_csv('pokedata.csv')
df_poke = df_poke.rename(make_lower, axis = 1)
df_poke = df_poke.set_index('pokedex')
df_poke.head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         attack  defense  
pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [23]:
# sort rows by index in decreasing order
df_poke.sort_index(ascending=False, axis=0).head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
pokedex                                                                        
151            Mew  Psychic           NaN      16     8.8     100        100   
150         Mewtwo  Psychic           NaN      79   269.0     106        130   
149      Dragonite   Dragon        Flying      87   463.0      91         80   
148      Dragonair   Dragon           NaN     157    36.4      61         70   
147        Dratini   Dragon           NaN      71     7.3      41         50   

         attack  defense  
pokedex                   
151         100      100  
150         110       90  
149         134       95  
148          84       65  
147          64       45

In [24]:
# sort columns by column names in increasing order (alphabetic)
df_poke.sort_index(axis=1).head()

attack  basespeed  damage  defense  height maintype     pokemon  \
pokedex                                                                    
1            49         45      45       49      24    Grass   Bulbasaur   
2            62         60      60       63      39    Grass     Ivysaur   
3            82         80      80       83      79    Grass    Venusaur   
4            52         65      39       43      24     Fire  Charmander   
5            64         80      58       58      43     Fire  Charmeleon   

        secondarytype  weight  
pokedex                        
1              Poison    25.2  
2              Poison    28.7  
3              Poison   220.5  
4                 NaN    18.7  
5                 NaN    41.9

In [25]:
# sort rows by a particular column value
df_poke.sort_values('weight').head()
#df_poke.sort_values('weight', ascending = False).head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
pokedex                                                                      
92        Gastly    Ghost        Poison      51    0.20      30         80   
93       Haunter    Ghost        Poison      63    0.20      45         95   
50       Diglett   Ground           NaN       8    1.80      10         95   
109      Koffing   Poison           NaN      24    2.00      40         35   
18       Pidgeot   Flying        Normal      59    2.21      83        101   

         attack  defense  
pokedex                   
92           35       30  
93           50       45  
50           55       25  
109          65       95  
18           80       75

In [26]:
# to sort columns with sort_values we must compare across columns
# this means all columns must have comparable types
# let's try reading in an excel doc *may need to install openpyxl
df = pd.read_excel('5by5.xlsx', index_col='index')
df

col0  col1  col2  col3  col4
index                              
row0      7     2     0     9     1
row1      3     3     2     9     9
row2      0     3     2     8     4
row3      1     2     2     3     7
row4      7     3     0     5     4

In [27]:
# sort columns so 'row0' is increasing across columns
df.sort_values(['row0', 'row1'], axis=1)
# df.sort_values(['row1', 'row2'], axis=1)

col2  col4  col1  col0  col3
index                              
row0      0     1     2     7     9
row1      2     9     3     3     9
row2      2     4     3     0     8
row3      2     7     2     1     3
row4      0     4     3     7     5

## Memory Management

By default, each of the methods above returns a **copy** of the dataframe. Instead of manually overwriting or creating a new data frame:

If you want to replace the dataframe with the output of some method, set `inplace=True` (available on many methods)

**Note**: this has the benefit of saving memory; instead of creating duplicates of a large data set, you update the data set itself. *However* you must be careful when making the changes, or you may have to re-start everything to get a clean data set.

In [28]:
df_poke = pd.read_csv('pokedata.csv')
df_poke = df_poke.rename(make_lower, axis = 1)
df_poke = df_poke.set_index('pokedex')
df_poke.head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         attack  defense  
pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [29]:
# this sort_index() returns a copy
df_poke.sort_index(axis=1).head()

attack  basespeed  damage  defense  height maintype     pokemon  \
pokedex                                                                    
1            49         45      45       49      24    Grass   Bulbasaur   
2            62         60      60       63      39    Grass     Ivysaur   
3            82         80      80       83      79    Grass    Venusaur   
4            52         65      39       43      24     Fire  Charmander   
5            64         80      58       58      43     Fire  Charmeleon   

        secondarytype  weight  
pokedex                        
1              Poison    25.2  
2              Poison    28.7  
3              Poison   220.5  
4                 NaN    18.7  
5                 NaN    41.9

In [30]:
# df_poke is the same as it ever was
df_poke.head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         attack  defense  
pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [31]:
# we can sort_index() `inplace` to avoid copying data
df_poke.sort_index(axis=1, inplace=True)
df_poke.head()

attack  basespeed  damage  defense  height maintype     pokemon  \
pokedex                                                                    
1            49         45      45       49      24    Grass   Bulbasaur   
2            62         60      60       63      39    Grass     Ivysaur   
3            82         80      80       83      79    Grass    Venusaur   
4            52         65      39       43      24     Fire  Charmander   
5            64         80      58       58      43     Fire  Charmeleon   

        secondarytype  weight  
pokedex                        
1              Poison    25.2  
2              Poison    28.7  
3              Poison   220.5  
4                 NaN    18.7  
5                 NaN    41.9

Common mistakes with `inplace`:
- assuming that a method/function works `inplace` when it doesn't
    - effect: your object hasn't been manipulated (though you wish it had been)
- assuming that a method/function returns a copy of object when it operates `inplace`
    - effect: your original data has been modified
        - symptom: your return value is `None`
        - symptom: your data appears to be changing 'by itself'

## Lecture Break/Practice

Load the `penguins` dataset from seaborn:

```python
import seaborn as sns

df_penguin = sns.load_dataset('penguins')
```

Manipulate `df_penguin` to produce the following DataFrame:

|        |         | beak depth (mm) | beak length (mm) | weight (g) | flipper (mm) |    sex |
|-------:|--------:|----------------:|-----------------:|-----------:|-------------:|-------:|
| **island** | **species** |                 |                  |            |              |        |
| Biscoe |  Adelie |            16.6 |             36.5 |     2850.0 |        181.0 | Female |
|        |  Adelie |            17.1 |             36.4 |     2850.0 |        184.0 | Female |
|        |  Adelie |            18.1 |             34.5 |     2900.0 |        187.0 | Female |
|        |  Adelie |            18.6 |             37.9 |     2925.0 |        193.0 | Female |
|        |  Adelie |            16.0 |             37.7 |     3075.0 |        183.0 | Female |
|        |  Adelie |            18.6 |             37.9 |     3150.0 |        172.0 | Female |
|        |  Adelie |            16.9 |             35.7 |     3150.0 |        185.0 | Female |

- practice being conservative in your use of memory by ensuring you only ever have a single copy of the data 
    - (i.e. modify it `inplace`)
- notice the dataframe above has a hierarchical index
- notice that the dataframe above is sorted per island, species and then weight
    - we need to sort by values in the index as well as values in the column ... did we cover this?
        - approach: before you `set_index()` to make a hierarchical index, set the row order with a `sort_values()`
- notice that the columns of the dataframe above has different column names too


When you're all done, find all the penguins on island `Dream` of species `Adelie` from this dataframe.  (The hierarchical index performs a similar function as boolean indexing.  Use the hierarchical index in the exercise please).

In [32]:
import numpy as np
import pandas as pd
import seaborn as sns

rename_dict = {'body_mass_g': 'weight (g)',
              'bill_depth_mm': 'beak depth (mm)',
              'bill_length_mm': 'beak length (mm)',
              'flipper_length_mm': 'flipper (mm)'}

# load data
df_penguin = sns.load_dataset('penguins')

df_penguin.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female

## Lecture Break/Practice Solution (available in day11_lecturebreaksol.ipynb)

In [33]:
# build the data frame

In [34]:
# grab the penguins we're interested in

## More DataFrame Manipulations

- missing values (NaN)
    - why they're there
    - what to do with them
        - discard data
        - fill all missing values in with something reasonable
    
- Duplicate rows
    - finding them
    - removing them

- Replacing values

- Applying functions to data
    - map()
    - applymap()
	- apply() 

## Missing Data

* Pandas automatically assigns **NaN**, "Not a Number", to missing values while reading a file
* some folks use other values to indicate a "missing" value
    - `-99` is popular, for some reason ...
    - use `na_values` argument of `read_csv()` to mark other values as **NaN**

In [35]:
# jupyter "magic" commands
# below shows the raw csv; notice that Charmander's missing SecondaryType is blank
!cat pokedata.csv

In [36]:
df_poke.head()

attack  basespeed  damage  defense  height maintype     pokemon  \
pokedex                                                                    
1            49         45      45       49      24    Grass   Bulbasaur   
2            62         60      60       63      39    Grass     Ivysaur   
3            82         80      80       83      79    Grass    Venusaur   
4            52         65      39       43      24     Fire  Charmander   
5            64         80      58       58      43     Fire  Charmeleon   

        secondarytype  weight  
pokedex                        
1              Poison    25.2  
2              Poison    28.7  
3              Poison   220.5  
4                 NaN    18.7  
5                 NaN    41.9

In [37]:
df_poke = pd.read_csv('pokedata.csv')
df_poke.rename(make_lower, axis = 1, inplace=True)
df_poke.set_index('pokedex', inplace=True)
df_poke.head()

pokemon maintype secondarytype  height  weight  damage  basespeed  \
pokedex                                                                         
1         Bulbasaur    Grass        Poison      24    25.2      45         45   
2           Ivysaur    Grass        Poison      39    28.7      60         60   
3          Venusaur    Grass        Poison      79   220.5      80         80   
4        Charmander     Fire           NaN      24    18.7      39         65   
5        Charmeleon     Fire           NaN      43    41.9      58         80   

         attack  defense  
pokedex                   
1            49       49  
2            62       63  
3            82       83  
4            52       43  
5            64       58

In [38]:
# what about data with som eother way of identifying missing data?
df_miss = pd.read_excel('miss_data.xlsx')
df_miss

patient height weight  trial0 trial1 trial2  trial3 trial4 finaltrial
0       a    180     78      45     39   miss      42     40       miss
1       b    164     67      47     49     43      48   miss       miss
2       c   miss   miss      42   miss     46      47     42       miss
3       d   miss   miss      23   miss   miss      17   miss       miss
4       e    199     89      38     40     44      48   miss       miss
5       f    134     54      34     36     29      22   miss       miss
6       g    163     59      50     50   miss      46     49       miss
7       h    175     72      43     37     35      31   miss       miss
8       i    158     80      28   miss     38      34     39       miss
9       j    169     79      33   miss     25      28   miss       miss

In [39]:
df_miss = pd.read_excel('miss_data.xlsx', index_col = 'patient', na_values = ('miss'))
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

### What to do with missing data?

* Option 0: Nothing
* Option 1: Discard all rows which are missing data
* Option 2: Fill in missing data with reasonble guesses (impute)



### What to do with missing data?
* Option 0: Nothing
    - When the `skipna=True`, we ignore NaN in our averaging
    - other pandas methods, besides `DataFrame.mean()`, have `skipna` too

In [40]:
df_miss = pd.read_excel('miss_data.xlsx', index_col = 'patient', na_values = ('miss'))
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [41]:
# mean of each column (skip any NaN values)
df_miss.mean(axis=0, skipna=True)

height        167.750000
weight         72.250000
trial0         38.300000
trial1         41.833333
trial2         37.142857
trial3         36.300000
trial4         42.500000
finaltrial           NaN
dtype: float64

In [42]:
# mean of each column (dont skip any NaN values)
df_miss.mean(axis=0, skipna=False)

height         NaN
weight         NaN
trial0        38.3
trial1         NaN
trial2         NaN
trial3        36.3
trial4         NaN
finaltrial     NaN
dtype: float64

In [43]:
# mean of each column (notice skipna=True by default)
df_miss.mean(axis=0)

height        167.750000
weight         72.250000
trial0         38.300000
trial1         41.833333
trial2         37.142857
trial3         36.300000
trial4         42.500000
finaltrial           NaN
dtype: float64

### What to do with missing data?

* Option 1: Discard all rows / columns which have missing data
    - appropriate when we have other data which contains similar information as discarded rows
    - `DataFrame.dropna()`: drop all rows or columns with missing data
        - axis
        - how: 'any', 'all' (do we drop row / column if ...
            - 'any' of its values are nan
            - 'all' of its values are nan

In [44]:
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [45]:
# drop rows with any nan values (default) ... nothing left
df_miss.dropna(axis=0, how='any')

Empty DataFrame
Columns: [height, weight, trial0, trial1, trial2, trial3, trial4, finaltrial]
Index: []

In [46]:
# drop columns with all nan values (unlikely, but possible)
df_miss.dropna(axis=1, how='all')

height  weight  trial0  trial1  trial2  trial3  trial4
patient                                                        
a         180.0    78.0      45    39.0     NaN      42    40.0
b         164.0    67.0      47    49.0    43.0      48     NaN
c           NaN     NaN      42     NaN    46.0      47    42.0
d           NaN     NaN      23     NaN     NaN      17     NaN
e         199.0    89.0      38    40.0    44.0      48     NaN
f         134.0    54.0      34    36.0    29.0      22     NaN
g         163.0    59.0      50    50.0     NaN      46    49.0
h         175.0    72.0      43    37.0    35.0      31     NaN
i         158.0    80.0      28     NaN    38.0      34    39.0
j         169.0    79.0      33     NaN    25.0      28     NaN

In [47]:
# drop row if it does not have at least three values that are **not** NaN
df_miss.dropna(thresh=3)
#df_miss.dropna(thresh=4)

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [48]:
# drop only if NaN in specific column
df_miss.dropna(subset=['trial4'])

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN

### What to do with missing data?

* Option 2: Fill in missing data
    - appropriate when we don't have much data, let's not waste any!
    - some known value
        - `DataFrame.fillna(value)`
    - some average value
        - `DataFrame.interpolate(method='zero')

In [49]:
# if the trial is missing, let us assume they deserve a 0
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [50]:
df_miss.fillna(0)

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     0.0      42    40.0         0.0
b         164.0    67.0      47    49.0    43.0      48     0.0         0.0
c           0.0     0.0      42     0.0    46.0      47    42.0         0.0
d           0.0     0.0      23     0.0     0.0      17     0.0         0.0
e         199.0    89.0      38    40.0    44.0      48     0.0         0.0
f         134.0    54.0      34    36.0    29.0      22     0.0         0.0
g         163.0    59.0      50    50.0     0.0      46    49.0         0.0
h         175.0    72.0      43    37.0    35.0      31     0.0         0.0
i         158.0    80.0      28     0.0    38.0      34    39.0         0.0
j         169.0    79.0      33     0.0    25.0      28     0.0         0.0

### Filling In a Different Value for Each Column

* Sometimes (as with height and weight) it doesn't make sense to give same value to everything
* Can fill in each column with a different value
* Pass in a dictionary containing column names as keys and desired NA values as values
    * **fillna(dictionary_of_values)**
    
    

In [51]:
dict_fill = {'height': 170,
             'weight': 70,
             'trial0': 50,
             'trial1': 45,
             'trial2': 40,
             'trial3': 35,
             'trial4': 30,
             'finaltrial': 25}
df_miss.fillna(dict_fill)

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0    40.0      42    40.0        25.0
b         164.0    67.0      47    49.0    43.0      48    30.0        25.0
c         170.0    70.0      42    45.0    46.0      47    42.0        25.0
d         170.0    70.0      23    45.0    40.0      17    30.0        25.0
e         199.0    89.0      38    40.0    44.0      48    30.0        25.0
f         134.0    54.0      34    36.0    29.0      22    30.0        25.0
g         163.0    59.0      50    50.0    40.0      46    49.0        25.0
h         175.0    72.0      43    37.0    35.0      31    30.0        25.0
i         158.0    80.0      28    45.0    38.0      34    39.0        25.0
j         169.0    79.0      33    45.0    25.0      28    30.0        25.0

### Interpoloation

https://en.wikipedia.org/wiki/2020_UK_GCSE_and_A-Level_grading_controversy

Because of the pandemic and the impossibility of organizing exams students got their imputed grades rather than take an exam.

How to implement that? We can fill in missing values with the average of the other row values

**But also, if you ever become a teacher, please don't do that to students.**

In [52]:
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [53]:
# Step 1: fill missing trials (not height or weight, or finaltrial) with 0
df_miss.fillna(value={'trial0': 0, 'trial1': 0, 'trial2': 0, 'trial3': 0, 'trial4': 0}, inplace=True)
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     0.0      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     0.0         NaN
c           NaN     NaN      42     0.0    46.0      47    42.0         NaN
d           NaN     NaN      23     0.0     0.0      17     0.0         NaN
e         199.0    89.0      38    40.0    44.0      48     0.0         NaN
f         134.0    54.0      34    36.0    29.0      22     0.0         NaN
g         163.0    59.0      50    50.0     0.0      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     0.0         NaN
i         158.0    80.0      28     0.0    38.0      34    39.0         NaN
j         169.0    79.0      33     0.0    25.0      28     0.0         NaN

In [54]:
# Step 2: get the mean score from first five trials per patient
means = df_miss[['trial0', 'trial1', 'trial2', 'trial3', 'trial4']].mean(axis=1)
means

patient
a    33.2
b    37.4
c    35.4
d     8.0
e    34.0
f    24.2
g    39.0
h    29.2
i    27.8
j    17.2
dtype: float64

In [55]:
# Step 3: Fill the missing values in finaltrial with the means
df_miss.fillna(value = {'finaltrial': means})

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     0.0      42    40.0        33.2
b         164.0    67.0      47    49.0    43.0      48     0.0        37.4
c           NaN     NaN      42     0.0    46.0      47    42.0        35.4
d           NaN     NaN      23     0.0     0.0      17     0.0         8.0
e         199.0    89.0      38    40.0    44.0      48     0.0        34.0
f         134.0    54.0      34    36.0    29.0      22     0.0        24.2
g         163.0    59.0      50    50.0     0.0      46    49.0        39.0
h         175.0    72.0      43    37.0    35.0      31     0.0        29.2
i         158.0    80.0      28     0.0    38.0      34    39.0        27.8
j         169.0    79.0      33     0.0    25.0      28     0.0        17.2

## Working with Duplicates

* Sometimes datasets contain duplicates record, which you may want to discard

### duplicated() method

Returns a boolean Series indicating whether each row is a duplicated (whether it has been observed in a previous row)

In [56]:
df_dupe = pd.read_excel('dupe_data.xlsx', na_values = ('miss'))
df_dupe

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0        a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1        b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2        c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3        d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4        e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5        f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
6        g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
7        g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
8        g   163.0    59.0      50    50.0    48.0      46    49.0         NaN
9        h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
10       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
11       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [57]:
# We have three rows with patient g
# row 6 and row 7 are identical, only row 7 is marked as duplicate (since row 6 is the original)

df_dupe.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7      True
8     False
9     False
10    False
11    False
dtype: bool

In [58]:
# we can specify the columns by which to tell dupes, this one finds all the duplicates of patient g
df_dupe.duplicated(subset='patient')

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7      True
8      True
9     False
10    False
11    False
dtype: bool

### drop_duplicates() method

* Returns a DataFrame where the duplicates have been dropped

In [59]:
df_dupe

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0        a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1        b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2        c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3        d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4        e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5        f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
6        g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
7        g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
8        g   163.0    59.0      50    50.0    48.0      46    49.0         NaN
9        h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
10       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
11       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [60]:
df_dupe.drop_duplicates()

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0        a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1        b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2        c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3        d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4        e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5        f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
6        g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
8        g   163.0    59.0      50    50.0    48.0      46    49.0         NaN
9        h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
10       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
11       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

* You can also specify a subset of columns to detect duplicates.
* Pass in a list of column names as an argument to the **drop_duplicates()** method call.

In [61]:
df_dupe.drop_duplicates(subset='patient')

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0        a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1        b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2        c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3        d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4        e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5        f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
6        g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
9        h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
10       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
11       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

* By default, both duplicated() and drop_duplicates() keep the first observed value combination. 
* If you want to keep the last value, pass in **keep="last"**
* This can be used for example where data is ordered chronologically and the last mention is the most recent one (here: patient g had a duplicate record because their trial 2 data was eventually found)

In [62]:
df_dupe.drop_duplicates(subset='patient', keep='last')

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0        a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1        b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2        c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3        d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4        e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5        f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
8        g   163.0    59.0      50    50.0    48.0      46    49.0         NaN
9        h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
10       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
11       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

## Replacing Values

* Beyond filtering or fill in missing data, we can replace certain values with new values.

### replace()

* **df.replace(current_value, new_value)**
* If you want to replace multiple values at once, pass a list and then the substitute value(s)


In [63]:
df = pd.DataFrame({'a': [2, 5, -99, 0, -999],
                   'b': [1, -999, 4, -99, 123]})
df

a    b
0    2    1
1    5 -999
2  -99    4
3    0  -99
4 -999  123

In [64]:
# replaces isntances of -99 or -999 in dataframe with NaN (across whole dataframe)
import numpy as np

df.replace([-99, -999], np.nan)

a      b
0  2.0    1.0
1  5.0    NaN
2  NaN    4.0
3  0.0    NaN
4  NaN  123.0

In [65]:
df

a    b
0    2    1
1    5 -999
2  -99    4
3    0  -99
4 -999  123

In [66]:
# replaces -99 with "NaN" and -999 with 0 (across whole dataframe)
df.replace([-99, -999], [np.nan, 0])
# equivalent to doing this twice in a row
# df = df.replace(-99, np.nan)
# df.replace(-999, 0)

a      b
0  2.0    1.0
1  5.0    0.0
2  NaN    4.0
3  0.0    NaN
4  0.0  123.0

In [67]:
# yet another way to replace -99 with "NaN" and -999 with 0 (across whole dataframe)
df.replace({-99: np.nan, -999: 0})

a      b
0  2.0    1.0
1  5.0    0.0
2  NaN    4.0
3  0.0    NaN
4  0.0  123.0

In [68]:
# replace -99 with "NaN" in column 'a'.  Replace -999 with 0 in column 'b'
df.replace({'a': -99, 'b': -999}, value={'a': np.nan, 'b': 0})

a    b
0    2.0    1
1    5.0    0
2    NaN    4
3    0.0  -99
4 -999.0  123

For example, all too often in datasets where the lat, lon is not known, the default value is 0 instead of NaN... but 0, 0 is an actual place in the world!

<img src="https://i.ibb.co/ZhT1Svb/google-maps.png" style="width: 60%;">

In [69]:
df_locs = pd.read_csv('locations.csv')
df_locs

city      lat       lon  population
0    Boston  42.3603  -71.0578      689326
1    Lusaka -15.4167   28.2833     2731696
2  Atlantis   0.0000    0.0000           0
3    Sydney -33.8678  151.2100     5231147

In [70]:
df_locs.replace(to_replace={'lat': 0, 'lon': 0}, value={'lat': np.nan, 'lon': np.nan})

city      lat       lon  population
0    Boston  42.3603  -71.0578      689326
1    Lusaka -15.4167   28.2833     2731696
2  Atlantis      NaN       NaN           0
3    Sydney -33.8678  151.2100     5231147

### map()  

* Perform a function on every element of a Series
* The function should return a value
* remember, each column in a DataFrame is a Series, so it can be performed on each value of a DataFrame column

In [71]:
df_miss = pd.read_excel('miss_data.xlsx', na_values = ('miss'))
df_miss

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0       a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1       b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2       c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3       d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4       e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5       f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
6       g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
7       h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
8       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
9       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [72]:
def introduce(pat_str):
    return f'Record for patient {pat_str}'

df_miss['patient'].map(introduce)

0    Record for patient a
1    Record for patient b
2    Record for patient c
3    Record for patient d
4    Record for patient e
5    Record for patient f
6    Record for patient g
7    Record for patient h
8    Record for patient i
9    Record for patient j
Name: patient, dtype: object

### applymap()

* Same as `map()` but rather with a Series, it runs the function elementwise on the whole DataFrame
* let's give every patient a bonus of 5 points

In [73]:
df_miss

patient  height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
0       a   180.0    78.0      45    39.0     NaN      42    40.0         NaN
1       b   164.0    67.0      47    49.0    43.0      48     NaN         NaN
2       c     NaN     NaN      42     NaN    46.0      47    42.0         NaN
3       d     NaN     NaN      23     NaN     NaN      17     NaN         NaN
4       e   199.0    89.0      38    40.0    44.0      48     NaN         NaN
5       f   134.0    54.0      34    36.0    29.0      22     NaN         NaN
6       g   163.0    59.0      50    50.0     NaN      46    49.0         NaN
7       h   175.0    72.0      43    37.0    35.0      31     NaN         NaN
8       i   158.0    80.0      28     NaN    38.0      34    39.0         NaN
9       j   169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [74]:
df_miss.set_index('patient', inplace=True)
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [75]:
df_miss_trials = df_miss.iloc[:,2:]
df_miss_trials

trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                    
a            45    39.0     NaN      42    40.0         NaN
b            47    49.0    43.0      48     NaN         NaN
c            42     NaN    46.0      47    42.0         NaN
d            23     NaN     NaN      17     NaN         NaN
e            38    40.0    44.0      48     NaN         NaN
f            34    36.0    29.0      22     NaN         NaN
g            50    50.0     NaN      46    49.0         NaN
h            43    37.0    35.0      31     NaN         NaN
i            28     NaN    38.0      34    39.0         NaN
j            33     NaN    25.0      28     NaN         NaN

In [76]:
def apply_bonus(x):
    return x+5

df_miss_trials.applymap(apply_bonus)

trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                    
a            50    44.0     NaN      47    45.0         NaN
b            52    54.0    48.0      53     NaN         NaN
c            47     NaN    51.0      52    47.0         NaN
d            28     NaN     NaN      22     NaN         NaN
e            43    45.0    49.0      53     NaN         NaN
f            39    41.0    34.0      27     NaN         NaN
g            55    55.0     NaN      51    54.0         NaN
h            48    42.0    40.0      36     NaN         NaN
i            33     NaN    43.0      39    44.0         NaN
j            38     NaN    30.0      33     NaN         NaN

### apply() method

* Apply a function along an axis of the DataFrame.
* The function should take a DataFrame and return either a Pandas object (DataFrame or Series) or a scalar

Let's calculate the mean of each patient's trials but without the worst score

In [77]:
df_miss

height  weight  trial0  trial1  trial2  trial3  trial4  finaltrial
patient                                                                    
a         180.0    78.0      45    39.0     NaN      42    40.0         NaN
b         164.0    67.0      47    49.0    43.0      48     NaN         NaN
c           NaN     NaN      42     NaN    46.0      47    42.0         NaN
d           NaN     NaN      23     NaN     NaN      17     NaN         NaN
e         199.0    89.0      38    40.0    44.0      48     NaN         NaN
f         134.0    54.0      34    36.0    29.0      22     NaN         NaN
g         163.0    59.0      50    50.0     NaN      46    49.0         NaN
h         175.0    72.0      43    37.0    35.0      31     NaN         NaN
i         158.0    80.0      28     NaN    38.0      34    39.0         NaN
j         169.0    79.0      33     NaN    25.0      28     NaN         NaN

In [78]:
def mean_skip_worst(row):
    # get only the trials, without the height and weight
    trials = row.loc['trial0':]
    
    # sort the values
    trials = trials.sort_values()
    
    # skip the worst score (first in the sorted series)
    best_trials = trials[1:]
    
    # return the mean
    return best_trials.mean()

In [79]:
# axis=1 => different inputs to function are different columns
# ! axis is kind of funny to remember here
df_miss.apply(mean_skip_worst, axis=1)

patient
a    42.333333
b    48.000000
c    45.000000
d    23.000000
e    44.000000
f    33.000000
g    49.666667
h    38.333333
i    37.000000
j    30.500000
dtype: float64

### In review:

- `map()` do item by item operation to a series
- `applymap()` do item by item operation to a whole dataframe
- `apply()` do row (or col) to item / row / col to a whole dataframe

## Lecture Break/Practice 2 (if time; if not, next time!)

* Load `grades.csv` treating -99 as missing values
* If any student `name` is mentioned multiple times, only keep the row with the last mention
* Assume that NaNs mean the student didn't submit, and replace them with 0
* use `apply()` to calculate the mean score of each assignment among all students without considering the both lowest *and* the highest score in each assignment 
    * you'll need to write a function which, given a list of values, returns the average of everything but the highest and lowest values

In [80]:
# load in the data

In [81]:
# replace NaN with 0

In [82]:
# get mean of quiz0 column, ignoring min and max, for practice

In [83]:
# define function that gets mean for any column

In [84]:
# apply to all columns

In [85]:
# (another way) how to do it without setting the index?